rfcs/202230110-graph-compiler/README.md

```diff
@@ -0,0 +1,224 @@
+ops, it marks the two nested loops in Tensor IR as “mergeable” during the
+lowering process. Then Tensor IR merges two nested loops mechanically as guided
+by the Graph IR optimizations.
+
+## Codegen
+
+Graph Compiler supports 3 different codegen for different purpose:
+
+- C codegen
+  - Can be used to generate offline C kernel
+  - Can be used for debugging purpose as the generated code is more friendly and
+    readable to developers.
+- LLVM codegen
+  - Provide best performance so far.
+  - Depends on LLVM project.
+- Builtin codegen
+  - Implemented with xbyak technology inside, no extra dependency.
+  - Supports AVX512 for now, will extend to AVX2.
+
+## Fusion Patterns
+
+For now, Graph compiler provides three major patterns:
+
+- The Multilayer Perceptron (MLP) workload contains multiple matmul ops
+  intermixed with activation ops like ReLU.
+- The Multi-Head Attention (MHA) subgraph, which is the key element to
+  Transformer based deep learning models like Bert for natural language
+  processing.
+- The convolution blocks subgraph, which contains 2-6 convolution layers
+  with activation, normalization and resiual_add.
+
+Each pattern provides below capabilites:
+- Cover both inference and training.
+- Cover fp32/bf16/int8 in inference and fp32/bf16 in training.
+- Cover AVX-512 and upper ISAs inlcuding AMX.
+
+## Performance
+
+We selected MLP inference and MHA inference as target workloads to evaluate
+the performance.
+
+![Matmul](images/mlp.png)
+
+![MHA](images/mha.png)
+
+## User Visible Options
+
+Below table contains the cmake option introduced by graph compiler:
+
+|CMake Option|Supported values (defaults in bold)|Description|
+|----| ----  | ----|
+|DNNL_GRAPH_BUILD_COMPILER_BACKEND|ON, **OFF**|Enables building graph compiler backend|
+|DNNL_GRAPH_COMPILER_JIT_BACKENDS|**llvm;c;builtin**|Select the codegen and JIT to build|
+|DNNL_GRAPH_LLVM_CONFIG|**AUTO**, {llvm-config EXECUTABLE}|Required when llvm codegen is selected. Defines the method for detecting/configuring LLVM|
+
+The environment variables introduced by graph compiler:
+
+Environment variables | Supported values (defaults in bold) | Description
+-- | -- | --
+ONEDNN_GRAPH_COMPILER_CPU_JIT | **llvm**, builtin, c | Specify the CPU codegen and JIT.
+ONEDNN_GRAPH_COMPILER_OPT_LEVEL | 0,1,2,**3** | Optimization level for both graph compiler and codegen.
+ONEDNN_GRAPH_COMPILER_KERNEL_TRACE | {**0**,1},{stderr,filename.json} | Generates kernel execution trace in given filename with chrome tracing format.
+ONEDNN_GRAPH_COMPILER_PRINT_PASS_RESULT | **0**, 1 | Print the output IR of each graph and tensor IR passes.
+ONEDNN_GRAPH_COMPILER_VERBOSE | **0**, 1, 2 | Print error or warning message during compilation.
+ONEDNN_GRAPH_COMPILER_DUMP_GENCODE | {path_to_dump} | Dump the genereated kernel in C
+ONEDNN_GRAPH_COMPILER_C_INCLUDE | {path_to_c_codegen_header} | Specify the C codegen header for JIT compliation.
```

Comment:
It's because our generated C kernel can't be built standalone. It requires some pre-defined headers in https://github.com/oneapi-src/oneDNN/tree/dev-graph/src/backend/graph_compiler/core/src/runtime/kernel_include. Those headers won't be included in the installation package. End-user need to specify the location by this environment if they want to use C codegen and JIT. We will provide more details in the documentation.